CMakePCHCompiler.cmake

```diff
@@ -70,43 +83,54 @@ function(target_precompiled_header) # target [...] header
 				message(WARNING "Unknown header type for language ${lang}")
 				set(header_type "c++-header")
 			endif()
-			if(MSVC)
-				# ensure pdb goes to the same location, otherwise we get C2859
-				file(TO_NATIVE_PATH
-					"${CMAKE_CURRENT_BINARY_DIR}${CMAKE_FILES_DIRECTORY}/${target}.dir"
-					pdb_dir
-					)
+			if(MSVC)			
 				# /Yc - create precompiled header
-				# /Fd - specify directory for pdb output
-				set(flags "/Yc /Fd${pdb_dir}\\")
+				# /Fp - exact location for precompiled header
+				# /FI - force include of precompiled header
+				set(flags "/Yc\"${win_header}\" /Fp\"${win_pch}\" /FI\"${win_header}\"")
+
+				#REMINDER OF EARLIER TRIBULATIONS
+				#The Microsoft.Cpp.Win32.targets script deletes the PCH file from 2010 forward.
+				#/Z7 is much simpler anyway, and I'm not sure what its real disadvantages are.
+				##set(debug_flags " /Fd\"${pdb_dir}\\\\\"")
+				#CMAKE_<lang>_FLAGS_DEBUG must be used. It's set at the top of the script.
+				#set(debug_flags " /Z7")
+				
+				set_source_files_properties(${header} PROPERTIES
+					LANGUAGE ${lang}
+					COMPILE_FLAGS ${flags})
+
+				add_library(${pch_target} OBJECT ${header})
+
 			else()
 				set(flags "-x ${header_type}")
+				set_source_files_properties(
+					${header}
+					PROPERTIES
+					LANGUAGE ${lang}PCH
+					COMPILE_FLAGS ${flags}
+					)
+				add_library(${pch_target} OBJECT ${header})
 			endif()
-			set_source_files_properties(
-				${header}
-				PROPERTIES
-				LANGUAGE ${lang}PCH
-				COMPILE_FLAGS ${flags}
-				)
-			add_library(${target}.pch OBJECT ${header})
-			set(pch_target ${target}.pch)
 		endif()
-		add_dependencies(${target} ${pch_target})
-		set(target_dir
-			${CMAKE_CURRENT_BINARY_DIR}${CMAKE_FILES_DIRECTORY}/${pch_target}.dir
-			)
-		if(MSVC)
-			get_filename_component(win_header "${header}" NAME)
-			file(TO_NATIVE_PATH "${target_dir}/${header}.pch" win_pch)
+
+		#set(exclude) to be removed later by __watch_pch_last_hook
+		if(MSVC)	
+			#Careful: set_target_properties is destructive
 			# /Yu - use given include as precompiled header
 			# /Fp - exact location for precompiled header
```

Comment:
What do you mean with destructive? That it overwrites existing compile flags? For that case set_properties provides an APPEND keyword.

Reply:
Basically different CMake directives that look very similar on the outside have subtle semantic differences. While I was working I found my settings being overwritten by this module, so I investigated and discovered that some CMake APIs/directives were better suited to modules than others (more courteous/conscientious.) 

IN SHORT the CMake directives the module uses currently have bad side effects that will no doubt become "Issues" in the future if they were not fixed here :)